pkg/mcs/resource_manager/server/server.go

```diff
@@ -0,0 +1,303 @@
+	}
+}
+
+// GetClient returns builtin etcd client.
+func (s *Server) GetClient() *clientv3.Client {
+	return s.etcdClient
+}
+
+// GetHTTPClient returns builtin http client.
+func (s *Server) GetHTTPClient() *http.Client {
+	return s.httpClient
+}
+
+// AddStartCallback adds a callback in the startServer phase.
+func (s *Server) AddStartCallback(callbacks ...func()) {
+	s.startCallbacks = append(s.startCallbacks, callbacks...)
+}
+
+// IsServing returns whether the server is the leader, if there is embedded etcd, or the primary otherwise.
+func (s *Server) IsServing() bool {
+	// TODO: implement this function with primary.
+	return atomic.LoadInt64(&s.isServing) == 1
+}
+
+// AddServiceReadyCallback adds the callback function when the server becomes the leader, if there is embedded etcd, or the primary otherwise.
+func (s *Server) AddServiceReadyCallback(callbacks ...func(context.Context)) {
+	s.primaryCallbacks = append(s.primaryCallbacks, callbacks...)
+}
+
+func (s *Server) initClient() error {
+	// TODO: We need to keep all backend endpoints and keep updating them to the latest. Once one of them failed, need to try another one.
+	tlsConfig, err := s.cfg.Security.ToTLSConfig()
+	if err != nil {
+		return err
+	}
+	endpoints := strings.Split(s.cfg.BackendEndpoints, ",")
+	for _, endpoint := range endpoints {
+		e, err := url.Parse(endpoint)
+		if err != nil {
+			return err
+		}
+		s.backendUrls = append(s.backendUrls, e)
+	}
+	if len(s.backendUrls) == 0 {
+		return errs.ErrURLParse.Wrap(errors.New("no backend url found"))
+	}
+	s.etcdClient, s.httpClient, err = etcdutil.CreateClients(tlsConfig, []url.URL{*s.backendUrls[0]})
+	return err
+}
+
+func (s *Server) startServer() error {
+	var mux cmux.CMux
+	tlsConfig, err := s.cfg.Security.ToTLSConfig()
+	if err != nil {
+		return err
+	}
+	if tlsConfig != nil {
+		l, err := tls.Listen(tcp, s.cfg.ListenAddr, tlsConfig)
+		if err != nil {
+			return err
+		}
+		mux = cmux.New(l)
+	} else {
+		l, err := net.Listen(tcp, s.cfg.ListenAddr)
+		if err != nil {
+			return err
+		}
+		mux = cmux.New(l)
+	}
+
+	grpcL := mux.MatchWithWriters(cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"))
+	httpL := mux.Match(cmux.Any())
+
+	manager := NewManager(s)
+	s.service = &Service{
+		ctx:     s.ctx,
+		manager: manager,
+	}
+
+	s.grpcServer = grpc.NewServer()
+	s.service.RegisterGRPCService(s.grpcServer)
+
+	handler, _ := SetUpRestHandler(s.service)
+	s.httpServer = &http.Server{
+		Handler:           handler,
+		ReadTimeout:       5 * time.Minute,
+		ReadHeaderTimeout: 5 * time.Second,
+	}
+
+	s.serverLoopWg.Add(2)
+	go func() {
+		defer s.serverLoopWg.Done()
+		s.grpcServer.Serve(grpcL)
+	}()
+	go func() {
+		defer s.serverLoopWg.Done()
+		s.httpServer.Serve(httpL)
+	}()
+
+	// Run callbacks
+	log.Info("triggering the start callback functions")
+	for _, cb := range s.startCallbacks {
+		cb()
+	}
+	// Server has started.
+	atomic.StoreInt64(&s.isServing, 1)
+	return mux.Serve()
```

Comment:
It seems that we should check mux.Serve() has no error then set "atomic.StoreInt64(&s.isServing, 1)" and declare  Server has started

Reply:
err = mux.Serve()
	if err == nil {
		// Server has started.
		atomic.StoreInt64(&s.isServing, 1)
	}

	return err

Reply:
feel this is the right logic

Reply:
mux.Serve() will block

Reply:
If the `Serve` has an error, do we need to change `isServing` back?

Reply:
yes, but if it runs normally, it will block and we cannot modify `isServing` later.

Reply:
thanks